### ISSUE TITLE
Feature request: additional $user variable containing an assignable LDAP field

### ISSUE THREAD

**Issue by mmattel** [owncloud/core#15427]
Based on #15313 

We have a oC user and, if enabled, a LDAP user.

Currently we assign the $user variable the oC logon user which can be used for eg SMB subfolder management.

It would be great to create an additional $user variant like `$user_ldap` with an definable LDAP field assigned to which can be used in the same way as $user.

Reason: The oC logon user name and the LDAP username used for subfolders on the target share may not be the same. 

Example attributes for AD:
http://www.selfadsi.de/user-attributes-w2k3.htm
http://www.selfadsi.de/user-attributes-w2k8.htm

**Comment by PVince81** [comment]
Assuming that you're talking about external storage

**Comment by Xenopathic** [comment]
How about being able to substitute an arbitrary variable from the Session object? Then user backends could pull data into the Session object. For example, `user_ldap` could pull a custom attribute into the Session as `my_custom1`, then during `setUserVars()` if `%{my_custom1}` is seen it gets that value from the Session and substitutes it in. If `%{key}` doesn't exist in the session it is replaced with the empty string.

**Comment by PVince81** [comment]
Possibly insecure CC @LukasReschke

**Comment by mmattel** [comment]
If we look into the documentation: https://github.com/owncloud/documentation/blob/master/admin_manual/configuration_user/user_auth_ldap.rst at section "Login Filter", we could reuse the idea. There is already a way where you assign the logon user for owncloud with an attribute from LDAP (other Attribute)
![image](https://cloud.githubusercontent.com/assets/3321281/7039567/1b67b54e-ddc4-11e4-9e66-ebccaeb80688.png)

Why not just adding another tab here where you assign in the same way ``$user_ldap`` ?

**Comment by blizzz** [comment]
Less settings, better. Typically you would use uid usually and samaccountname with AD. We discover this already for the login filter. Might be worth about adding a functionality to replace ownCloud user names with LDAP usersnames were necessary. This change, or rather addition, would reside on user backend level. But when you decide to replace it? Maybe external resources nevertheless listen to the ownCloud username for whatever reason? Then again you would need settings here and there again. However, for a good solution do not think alone of LDAP and files_external, because there is more than that. The old mozilla sync app, for example, also took LDAP email instead of the user name, but that was available already, but it would have benefit from a proper solution as well (to just give one example).

**Comment by mmattel** [comment]
I have to be honest, I do not use the ldap plugin...
But I use files_external with SMB and I listen to other users of owncloud.

Especially with files_external, the idea and availability to use the current logon user name as predefined subdirectory name ``$user`` is really a good feature.

It seems that users have in some circumstances the problem, that the logonuser derived from an ldap attribute, does not match the needs for a subdirectory name as mentioned. Thinking about that the logon username chosen in owncloud is the email address derived from ldap while the subdirectory name should be the ldap uid as this may already exist on the smb storage used.

My proposal was to find a solution for that issue. The point is, form my limited view, that $user can not be used as described above, so a new ``$user_ldap`` variable with an assignable ldap attribute is necessary.

I think that you have seen the issue, you are the expert, maybe you find an solution for that

**Comment by Xenopathic** [comment]
Does it ever make sense to use the ownCloud UID as the replacement for `$user`, or the login name? When using LDAP, the UID is usually set to some bizarre UUID which is completely unusable, but the login name is the real LDAP username (which is probably used on all other services connected to LDAP). When designing SMB_OC, I specifically used the session login name instead of the UID, due to this exact problem. I wonder if the easiest solution is to simply change what `$user` is substituted with to the login name, rather than the UID.

**Comment by mmattel** [comment]
I have created a small Powerpoint having three slides.
Here are the screenshots to explain the problem in more detail:

**Slide 1**
![image](https://cloud.githubusercontent.com/assets/3321281/7101362/e10c23a4-e056-11e4-914a-33603b75f7d3.png)

**Slide 2**
![image](https://cloud.githubusercontent.com/assets/3321281/7101376/71d0bd78-e057-11e4-807e-aab3bf1e38b8.png)


**Slide 3**
![image](https://cloud.githubusercontent.com/assets/3321281/7101383/bc68238a-e057-11e4-8387-66dd88b1ece4.png)

I wanted to show with ``$user_ldap`` a way, which would definitely work. Maybe there is another solution possible, but the requirement for a solution is here.

**Comment by Xenopathic** [comment]
We really have to avoid linking files_external with user_ldap. Those sort of dependencies are difficult to manage, and are easily broken. Any information needed by files_external comes from objects or interfaces in core oC, which user_ldap can populate as necessary.

**Comment by mmattel** [comment]
> Any information needed by files_external comes from objects or interfaces in core oC, which user_ldap can populate as necessary.

Please describe how an admin can define in field files_external/SMB/subfolder a particular user_ldap attribute which is then further used by onwcloud in the same way or functionality as $user is implemented right now

I am absolutely open to any idea that solves the issue I have described it in detail in the three screenshots above. As said, I do not stick to a new variable, but we need a functionality matching the requirement.

**Comment by Xenopathic** [comment]
That is why I asked before, is does it make sense for `$user` to be an unusable field?

> Does it ever make sense to use the ownCloud UID as the replacement for $user, or the login name? When using LDAP, the UID is usually set to some bizarre UUID which is completely unusable, but the login name is the real LDAP username (which is probably used on all other services connected to LDAP). When designing SMB_OC, I specifically used the session login name instead of the UID, due to this exact problem. I wonder if the easiest solution is to simply change what $user is substituted with to the login name, rather than the UID.

**Comment by mmattel** [comment]
@Xenopathic 
There are two different things we should not mix up

Info: I am using the AD LDAP attribute names in the following examples

1. Credentials used for mounting a share
2. Restricting the user to use only a particular subfolder in the share (his personal root)

ad 1: The difference in having SMB and SMB_OC is, that in SMB an admin ususally uses admin credentials the user does not know to access the share and provide this share then to users while in SMB_OC, the crentials of the logged on user is used to access the share. This is also reflected in the missing user/pwd fields there. But this is not what we are talking here.

ad 2: When accessing a share, the logged on user can be enforced by onwcloud to start at a particular subdirectory in the share. This is the so called homedir functionality. In AD you have \\\server\share\%username% for this functionality. While admins can access by their granted rights the whole share as a seperate mount, users can not see the whole thing - only their personal one. ``%username%`` is in AD substituted by the login name which is the LDAP attribute ``sAMAaccountName``

owncloud populates a variable called ``$user`` with the logon name used. This variable can be entered  in the ``subdirectory`` field when defining SMB in the GUI. In mount.json it has the name ``root``.
Having that, the user is dircted by onwcloud to the particular subdirectory having the logon user name.

Up to here everything is fine.

For the particular case, that LDAP is used AND the admin decides NOT to use the LDAP attribute sAMAaccountName but as described in https://github.com/owncloud/documentation/blob/master/admin_manual/configuration_user/user_auth_ldap.rst but --> Logon Filter --> Edit raw filter instead --> username or email address: (|(uid=%uid)(mail=$uid))
this functionality breakes, because when the eMail address is used to logon, $user is populated by owncloud with eg name@host.com because the user choses the eMail address for logging on. Following the substitution of $user for the subdirectory, owncloud will try to access a subdirectory having as name the eMail address which is not present at the share given.

As the functionality $user is a good mechanism, my idea was to add a second variable with a particular name with the same functionality but where the admin can define which substituion from LDAP will be used. Again, any idea that makes that requirement working is welcomed, having a additional variable is maybe one way out of many possible.

**Comment by blizzz** [comment]
> Does it ever make sense to use the ownCloud UID as the replacement for $user, or the login name? When using LDAP, the UID is usually set to some bizarre UUID which is completely unusable, but the login name is the real LDAP username (which is probably used on all other services connected to LDAP).

Specifially to files_external you have a point here. While, hypothetically the file storage, whatever it may be, could check against ownCloud users.

> When designing SMB_OC, I specifically used the session login name instead of the UID, due to this exact problem. I wonder if the easiest solution is to simply change what $user is substituted with to the login name, rather than the UID.

Might work. Not sure there is a scenario where files would be accessed before the first login i.e. unknown login name.

cc @icewind1991

**Comment by sheepherd** [comment]
I've run into this issue too with OC 9.0.1.
Is is possible yet to set a custom $user variable?

**Comment by blizzz** [comment]
nothing changed in this regard.

**Comment by evil79genius** [comment]
I'm evaluating OC9 right now: facing the same issue when integrating with AD, I've found this thread.
I've solved it configuring LDAP, "Expert" tab, and setting "Override UUID detection / UUID Attribute for Users" to sAMAccountName, then clearing Username-LDAP user mapping with the button below.

One warning: doing so will "reset" user's data on the OC server - well, it will not reset it, it's just that the user will seem like a different one to the system.
